src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sstring.hh>
+
+#include <memory>
+
+namespace ss = seastar;
+
+namespace util {
+
+/// Tracks allocations and deallocations of memory. A mem_tracker can have
+/// child trackers effectively forming an n-ary tree like structure.
+class mem_tracker {
+public:
+    explicit mem_tracker(ss::sstring label)
+      : _label(std::move(label)) {}
+    mem_tracker(mem_tracker&) = delete;
+    mem_tracker(mem_tracker&&) = default;
+    mem_tracker& operator=(mem_tracker&) = delete;
+    mem_tracker& operator=(mem_tracker&&) = delete;
+    ~mem_tracker() = default;
+
+    void allocate(int64_t bytes) { _consumption += bytes; }
+
+    void deallocate(int64_t bytes) { _consumption -= bytes; }
+
+    ss::lw_shared_ptr<mem_tracker> create_child(ss::sstring);
+
+    /// Consumption of self + all the children rooted with this tracker..
+    int64_t consumption() const;
+
+    /// Pretty prints the tracker tree rooted at this instance.
+    /// Example:
+    /// |
+    /// root: 32.000bytes
+    /// | |
+    /// | child1: 24.000bytes
+    /// | | |
+    /// | | child11: 0.000bytes
+    /// | | | |
+    /// | | | child111: 0.000bytes
+    /// | |
+    /// | child2: 0.000bytes
+    ss::sstring debug(int level = 0) const;
+
+private:
+    ss::sstring _label;
+    int64_t _consumption = 0;
+    intrusive_list_hook _hook;
+    intrusive_list<mem_tracker, &mem_tracker::_hook> _children;
+};
+
+/// An allocator paired with a memory tracker to account for consumption.
+template<class T, class allocator = std::allocator<T>>
+class tracking_allocator : public allocator {
+public:
+    explicit tracking_allocator(ss::lw_shared_ptr<mem_tracker> tracker)
+      : _tracker(tracker) {}
+
+    template<class U>
+    explicit tracking_allocator(const tracking_allocator<U>& other)
+      : allocator(other)
+      , _tracker(other.tracker()) {}
+
+    template<class U>
+    struct rebind {
+        using other = tracking_allocator<U, std::allocator<U>>;
+    };
+
+    [[nodiscard]] T* allocate(std::size_t n) {
+        _tracker->allocate(n * sizeof(T));
+        return allocator::allocate(n);
+    }
+
+    void deallocate(T* p, std::size_t n) {
+        allocator::deallocate(p, n);
+        _tracker->deallocate(n * sizeof(T));
+    }
+
+    ss::lw_shared_ptr<mem_tracker> tracker() const { return _tracker; }
+
+private:
+    ss::lw_shared_ptr<mem_tracker> _tracker;
```

Comment:
A major concern I have is that this changes any container which uses it from a "type 2" object (i.e., if you copy an object the original and copy can be use concurrently on different threads) to a "type 3" object (even distinct objects cannot necessarily be used safely across threads, because they many internally share state which is unsafe for concurrent modification). 

This means the containers are not really as general purpose as the originals since we have to be careful never to pass them across shards. Maybe it's still OK if we very carefully document this? Another option would be to make this all thread-safe, e.g., with `std::shared_ptr` as the pointer type and atomics for the counters.